Dynamic virtual populate not working on subdocument with separate schema

### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported


### Mongoose version

6.5.4

### Node.js version

16.15.0

### MongoDB server version

4.2.1

### Description

Seems to be related to #8277 and #8742

When using dynamic ref on child schema `{ ref: (doc) => doc.refType }`, the passed in `doc` is the parent document, not the subdocument, causing dynamic ref path not correctly returned.

I tried both `virtuals` option in schema constructor options and schema `virtual()` method, tried both `ref` and `refPath` in virtual options, none works.

### Steps to Reproduce

```typescript
import mongoose from 'mongoose';
import { expect } from 'chai';

const nestedSchema = new mongoose.Schema({
    targetType: String,
    targetId: mongoose.Schema.Types.ObjectId,
}, {
    virtuals: {
        target: {
            options: {
                ref: (doc: any) => {
                    // doc should be the sub document
                    // but instead, it is the parent document
                    console.debug(doc);
                    return doc.targetType;
                },
                localField: 'targetId',
                foreignField: '_id',
                justOne: true,
            }
        }
    }
});

const parentSchema = new mongoose.Schema({
    nested: {
        type: nestedSchema
    },
});

const NestedDynamicVirtualPopulateTest = mongoose.model('NestedDynamicVirtualPopulateTest', parentSchema);

await NestedDynamicVirtualPopulateTest.collection.drop();
const target = new NestedDynamicVirtualPopulateTest({});
await target.save();

const parent = new NestedDynamicVirtualPopulateTest({
    nested: {
        targetType: (target.constructor as typeof NestedDynamicVirtualPopulateTest).modelName,
        targetId: target._id,
    }
});
await parent.save();
await parent.populate('nested.target');
// @ts-ignore
expect(parent.nested.target).to.exist;
```

### Expected Behavior

Since the virtual is defined on the child schema, the dynamic ref path should be relative to the subdocument, not the parent document.

![image](https://user-images.githubusercontent.com/39510674/187711998-bf118b26-2119-4330-bbde-29f35b9c51af.png)

```
12363.ts:49:29 - error TS2551: Property 'target' does not exist on type '{ targetType?: {}; targetId?: { toString: any; _bsontype?: { toString: {}; _bsontype?: ObjectId; inspect?: {}; equals?: {}; _id?: ...; id?: { [x: number]: unknown; [toStringTag]?: unknown; [iterator]?: {}; toString: {}; ... 100 more ...; byteOffset?: unknown; }; toHexString?: {}; toJSON?: {}; generationTime?: unknow...'. Did you mean 'targetId'?

49   console.log(parent.nested.target);
                               ~~~~~~

node_modules/mongoose/types/query.d.ts:619:29 - error TS2304: Cannot find name 'this'.

619     toConstructor(): typeof this;
                                ~~~~


Found 2 errors.
```


```typescript
import mongoose from 'mongoose';

const nestedSchema = new mongoose.Schema({
    targetType: String,
    targetId: mongoose.Schema.Types.ObjectId,
}, {
    virtuals: {
        target: {
            options: {
                ref: (doc: any) => {
                    // doc should be the sub document
                    // but instead, it is the parent document
                    console.debug(doc);
                    return doc.targetType;
                },
                localField: 'targetId',
                foreignField: '_id',
                justOne: true,
            }
        }
    }
});

const parentSchema = new mongoose.Schema({
    nested: {
        type: nestedSchema
    },
});

const NestedDynamicVirtualPopulateTest = mongoose.model('NestedDynamicVirtualPopulateTest', parentSchema);
async function run() {
  await mongoose.connect('mongodb://localhost:27017');
  await mongoose.connection.dropDatabase();
  const target = new NestedDynamicVirtualPopulateTest({});
  await target.save();

  const parent = new NestedDynamicVirtualPopulateTest({
      nested: {
          targetType: (target.constructor as typeof NestedDynamicVirtualPopulateTest).modelName,
          targetId: target._id,
      }
  });
  await parent.save();
  await parent.populate('nested.target');
  console.log(parent);
  console.log('=============');
  console.log(parent.nested);
  console.log('=================');
  console.log(parent.nested.target);
  console.log('======================')
}

run();
```

--- a/lib/helpers/populate/getModelsMapForPopulate.js
+++ b/lib/helpers/populate/getModelsMapForPopulate.js
@@ -410,7 +410,15 @@ function _virtualPopulate(model, docs, options, _virtualRes) {
       justOne = options.justOne;
     }
 
-    modelNames = virtual._getModelNamesForPopulate(doc);
+    // Use the correct target doc/sub-doc for dynamic ref on nested schema. See gh-12363
+    if (_virtualRes.nestedSchemaPath && typeof virtual.options.ref === 'function') {
+      const subdocs = utils.getValue(_virtualRes.nestedSchemaPath, doc);
+      modelNames = Array.isArray(subdocs)
+        ? subdocs.flatMap(subdoc => virtual._getModelNamesForPopulate(subdoc))
+        : virtual._getModelNamesForPopulate(subdocs);
+    } else {
+      modelNames = virtual._getModelNamesForPopulate(doc);
+    }
     if (virtual.options.refPath) {
       justOne = !!virtual.options.justOne;
       data.isRefPath = true;

--- a/test/model.populate.test.js
+++ b/test/model.populate.test.js
@@ -4280,10 +4280,11 @@ describe('model: populate:', function() {
         ]);
         await Parent.create([
           { b: { name: 'test1', referencedModel: 'Test1', aId: as[0]._id } },
-          { b: { name: 'test2', referencedModel: 'Test2', aId: as[1]._id } }
+          { b: { name: 'test2', referencedModel: 'Test2', aId: as[1]._id } },
+          { b: { name: 'test3', referencedModel: 'Test2', aId: '0'.repeat(24) } }
         ]);
         const parents = await Parent.find().populate('b.a').sort({ _id: 1 });
-        assert.deepStrictEqual(parents.map(p => p.b.a.name), ['a1', 'a2']);
+        assert.deepStrictEqual(parents.map(p => p.b.a?.name), ['a1', 'a2', undefined]);
       });
 
       it('with functions for match (gh-7397)', async function() {

--- a/test/model.populate.test.js
+++ b/test/model.populate.test.js
@@ -4247,6 +4247,45 @@ describe('model: populate:', function() {
           catch(done);
       });
 
+      it('with functions for ref with subdoc virtual populate (gh-12440) (gh-12363)', async function() {
+        const ASchema = new Schema({
+          name: String
+        });
+
+        const BSchema = new Schema({
+          referencedModel: String,
+          aId: ObjectId
+        });
+
+        BSchema.virtual('a', {
+          ref: function() {
+            return this.referencedModel;
+          },
+          localField: 'aId',
+          foreignField: '_id',
+          justOne: true
+        });
+
+        const ParentSchema = new Schema({
+          b: BSchema
+        });
+
+        const A1 = db.model('Test1', ASchema);
+        const A2 = db.model('Test2', ASchema);
+        const Parent = db.model('Parent', ParentSchema);
+
+        const as = await Promise.all([
+          A1.create({ name: 'a1' }),
+          A2.create({ name: 'a2' })
+        ]);
+        await Parent.create([
+          { b: { name: 'test1', referencedModel: 'Test1', aId: as[0]._id } },
+          { b: { name: 'test2', referencedModel: 'Test2', aId: as[1]._id } }
+        ]);
+        const parents = await Parent.find().populate('b.a').sort({ _id: 1 });
+        assert.deepStrictEqual(parents.map(p => p.b.a.name), ['a1', 'a2']);
+      });
+
       it('with functions for match (gh-7397)', async function() {
         const ASchema = new Schema({
           name: String,
@@ -6642,8 +6681,8 @@ describe('model: populate:', function() {
       });
 
       clickedSchema.virtual('users_$', {
-        ref: function(doc) {
-          return doc.events[0].users[0].refKey;
+        ref: function(subdoc) {
+          return subdoc.users[0].refKey;
         },
         localField: 'users.ID',
         foreignField: 'employeeId'
@@ -6706,8 +6745,8 @@ describe('model: populate:', function() {
       });
 
       clickedSchema.virtual('users_$', {
-        ref: function(doc) {
-          const refKeys = doc.events[0].users.map(user => user.refKey);
+        ref: function(subdoc) {
+          const refKeys = subdoc.users.map(user => user.refKey);
           return refKeys;
         },
         localField: 'users.ID',

